### ISSUE TITLE
Feature Request: Indexable writes to netCDF format files on disk

### ISSUE THREAD

**Issue by songololo** [pydata/xarray#2823]
This is a feature request for indexable writes to a netCDF format file.

The documents mention that data variables (my interpretation is: the entire multi-dimensional array represented as a data-variable entity) can be appended-to or overwritten when using `ds.to_netcdf(<filepath>)`. However, this requires first reading the entire dataset into memory, then modifying some-or-other element at some-or-other index in a multi-dimensional array, and then replacing the entire data-variable.

It would be incredibly useful to be able to write a single data value to a single array location without first loading and then completely overwriting the complete data-variable.

There may be ways to achieve a similar result by using the underlying `netCDF4` module, e.g. per [this stack overflow post](https://stackoverflow.com/a/46957861/1190200). But forwards-compatibility and unintended side-effects put a damper on such approaches.

**Comment by shoyer** [comment]
Yes, I think this would be a welcome feature, assuming we can figure out a satisfactory API.

Likely there's some similarity to the more specific task of appending along an existing dimension (https://github.com/pydata/xarray/issues/1672).
